### PR TITLE
Users/jasholl/jenkins queue job endpoint

### PR DESCRIPTION
**Pull request by jasholl**
This is not yet ready for merge.

### CODE REVIEW

File: Tasks/JenkinsQueueJob/util.ts [hunk 1]
```diff
@@ -115,40 +115,92 @@ function createRootJob(queueUri: string, jobQueue: JobQueue, taskOptions: TaskOp
...
     var defer: Q.Deferred<string> = Q.defer<string>();
-    tl.debug('submitJob(): ' + JSON.stringify(initialPostData));
+    tl.debug('submitJob(): ' + JSON.stringify(taskOptions));
 
-    request.post(initialPostData, function requestCallback(err, httpResponse, body) {
-        tl.debug('submitJob().requestCallback()');
+    function addCrumb(postData: any): any {
+        if (taskOptions.crumb && taskOptions.crumb != taskOptions.NO_CRUMB) {
+            postData.headers = {
+                'Jenkins-Crumb': taskOptions.crumb
```
Comment (AlexRukhlin) [inline Tasks/JenkinsQueueJob/util.ts]: You probably shouldn't hardcode the crumb header name. The name is provided in the response to the crumbIssuer request. to

File: Tasks/JenkinsQueueJob/util.ts [hunk 2]
```diff
@@ -158,4 +210,113 @@ function submitJob(initialPostData: any, taskOptions: TaskOptions): Q.Promise<st
...
+            }
+        } else if (httpResponse.statusCode == 404) {
+            tl.debug('crumb not found');
+            taskOptions.crumb = taskOptions.NO_CRUMB;
+            defer.resolve(taskOptions.NO_CRUMB);
+        } else if (httpResponse.statusCode != 200) {
+            failReturnCode(httpResponse, 'crumb request failed.');
+            defer.reject(getFullErrorMessage(httpResponse, 'Crumb request failed.'));
+        } else {
+            taskOptions.crumb = body.slice(body.indexOf(':') + 1);
```
Comment (AlexRukhlin) [inline Tasks/JenkinsQueueJob/util.ts]: You ignore the provided header name and use the hard coded one. :-(

File: Tasks/JenkinsQueueJob/util.ts [hunk 3]
```diff
@@ -158,4 +210,113 @@ function submitJob(initialPostData: any, taskOptions: TaskOptions): Q.Promise<st
...
+        }
+        var paramName = paramLine.substr(0, splitIndex).trim();
+        var paramValue = paramLine.slice(splitIndex + 1).trim();
+        formData[paramName] = paramValue;
+    }
+    return formData;
+}
+
+//https://www.visualstudio.com/docs/build/define/variables
+var allTeamBuildVariables: string[] = [
```
Comment (AlexRukhlin) [inline Tasks/JenkinsQueueJob/util.ts]: What if the list of variables changes? (AFAIK it changes from time to time.)

File: Tasks/JenkinsQueueJob/util.ts [hunk 4]
```diff
@@ -115,47 +115,219 @@ function createRootJob(queueUri: string, jobQueue: JobQueue, taskOptions: TaskOp
...
+    'Build.RequestedForId',
+    'Build.SourceBranch',
+    'Build.SourceBranchName',
+    'Build.SourcesDirectory',
+    'Build.SourceVersion',
+    'Build.StagingDirectory',
+    'Build.Repository.Git.SubmoduleCheckout',
+    'Build.SourceTfvcShelveset',
+    'Common.TestResultsDirectory',
+    'System.AccessToken',
```
Comment (DavidStaheli) [inline Tasks/JenkinsQueueJob/util.ts]: We shouldn't pass the OAuth access token to Jenkins (`System.AccessToken`).  Jenkins has its own credentials for connecting to TFS/Team Services.

File: Tasks/JenkinsQueueJob/util.ts (hunk 1 above)
Comment (jasholl) [inline Tasks/JenkinsQueueJob/util.ts]: fixed, thanks

File: Tasks/JenkinsQueueJob/util.ts (hunk 2 above)
Comment (jasholl) [inline Tasks/JenkinsQueueJob/util.ts]: fixed, thanks

File: Tasks/JenkinsQueueJob/util.ts (hunk 3 above)
Comment (jasholl) [inline Tasks/JenkinsQueueJob/util.ts]: To my knowledge, there is no mechanism in place right now to get all of the variables.  We'll need to update this if there are new variables that come along we need to pass.

File: Tasks/JenkinsQueueJob/util.ts (hunk 4 above)
Comment (jasholl) [inline Tasks/JenkinsQueueJob/util.ts]: fixed.

**Pull request merged**